shared/src/types/ibc.rs

```diff
@@ -188,3 +191,422 @@ impl ClientUpgradeData {
             .map_err(|e| Error::DecodingError(e.to_string()))
     }
 }
+
+/// Data to initialize a connection
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize)]
+pub struct ConnectionOpenInitData {
+    /// The corresponding client ID
+    client_id: String,
+    /// The corresponding counterparty
+    counterparty: Vec<u8>,
+    /// The version
+    version: Vec<u8>,
+    /// The delay period as (secs, nanos)
+    delay_period: DurationNanos,
+}
+
+impl ConnectionOpenInitData {
+    /// Returns the data to initalize a connection
+    pub fn new(
+        client_id: ClientId,
+        counterparty: Counterparty,
+        version: Version,
+        delay_period: Duration,
+    ) -> Self {
+        let client_id = client_id.as_str().to_owned();
+        // TODO: Need Profobuf implementation for Counterparty in ibc-rs
+        // let counterparty = counterparty.encode_vec().expect("Encoding a
+        // counterparty shouldn't fail");
+        let mut bytes = vec![];
+        RawCounterparty::from(counterparty)
+            .encode(&mut bytes)
+            .expect("Encoding a counterparty shouldn't fail");
+        let version = version
+            .encode_vec()
+            .expect("Encoding a version shouldn't fail");
+        Self {
+            client_id,
+            counterparty: bytes,
+            version,
+            delay_period: delay_period.into(),
+        }
+    }
+
+    /// Returns the client ID
+    pub fn client_id(&self) -> Option<ClientId> {
+        ClientId::from_str(&self.client_id).ok()
+    }
+
+    /// Returns the counterparty
+    pub fn counterparty(&self) -> Option<Counterparty> {
+        // TODO: Need Profobuf implementation for Counterparty in ibc-rs
+        // Counterparty::decode_vec(self.counterparty).ok()
+        match RawCounterparty::decode(&self.counterparty[..]) {
+            Ok(c) => c.try_into().ok(),
+            Err(_) => None,
+        }
+    }
+
+    /// Returns the version
+    pub fn version(&self) -> Option<Version> {
+        Version::decode_vec(&self.version).ok()
+    }
+
+    /// Returns the delay period
+    pub fn delay_period(&self) -> Duration {
+        Duration::new(self.delay_period.secs, self.delay_period.nanos)
+    }
+}
+
+/// Data to try to open a connection
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize)]
+pub struct ConnectionOpenTryData {
+    /// The previous connection ID
+    prev_conn_id: Option<String>,
```

Comment:
why do these fields need to be provided in the message data, aren't they inferable from the state writes?

Reply:
I removed `prev_conn_id`. And, I modified validations to use connection info instead of these data because we have to check the created/updated connection is valid. (The connection should be created/updated according to these data.)
I think others are used in a transaction to create/update connections.